### PR TITLE
Bug 1898954: Backup script should get the latest revision of static-pod-resources …

### PR DESCRIPTION
**Pull request by retroflexer**
…in a reliable way.

Currently the backup script relies on the timestamp on the filesystem to determine the latest revision of the static pod resources. Instead, we should use more reliable ways such as inspecting the manifests for the currently running containers to exactly determine the currently active revision.

### CODE REVIEW

File: bindata/etcd/cluster-backup.sh [hunk 1]
```diff
@@ -34,10 +34,9 @@ function backup_latest_kube_static_resources {
 
   LATEST_RESOURCE_DIRS=()
   for RESOURCE in "${RESOURCES[@]}"; do
-    # shellcheck disable=SC2012
-    LATEST_RESOURCE=$(ls -trd "${CONFIG_FILE_DIR}"/static-pod-resources/"${RESOURCE}"-[0-9]* | tail -1) || true
+    LATEST_RESOURCE=$(crictl inspect "$(crictl ps --name "${RESOURCE}"$ -q)" | jq -r '.status.mounts | .[] | select (.containerPath=="/etc/kubernetes/static-pod-resources") | .hostPath') || true
     if [ -z "$LATEST_RESOURCE" ]; then
-      echo "error finding static-pod-resource ${RESOURCE}"
+      echo "error finding static-pod-resource for the ${RESOURCE} pod"
```
Comment (hexfusion) [inline bindata/etcd/cluster-backup.sh:39]: small nit maybe we should add "please check if it is running" here? That seems to be the primary failure for crictl to error because the container is maybe crash looping thus not running.

**Review by hexfusion — COMMENTED** [review]
overall the approach seems reasonable, please confirm jq status for OCP 4.4+ to ensure there is no corner cases.

**Comment by marun** [comment]
Can you confirm whether the disruptive test failure is related to the proposed change?

**Comment by retroflexer** [comment]
> Can you confirm whether the disruptive test failure is related to the proposed change?

The error looks like some quota limit, and doesn't seem to have anything to do with the change in the script:
```
 level=error msg=Error: Error creating VPC: VpcLimitExceeded: The maximum number of VPCs has been reached
```

**Comment by marun** [comment]
> > Can you confirm whether the disruptive test failure is related to the proposed change?
> 
> The error looks like some quota limit, and doesn't seem to have anything to do with the change in the script:
> 
> ```
>  level=error msg=Error: Error creating VPC: VpcLimitExceeded: The maximum number of VPCs has been reached
> ```

https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/504/pull-ci-openshift-cluster-etcd-operator-master-e2e-disruptive/1334306435525251072

**Comment by retroflexer** [comment]
> https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/504/pull-ci-openshift-cluster-etcd-operator-master-e2e-disruptive/1334306435525251072

This change assumes that all the static pods are running and can be queried using `crictl inspect` at the time of backup. Does your test somehow not have the static pods running?

**Comment by retroflexer** [comment]
Adding `set -x` to debug the code.

**Comment by marun** [comment]
> > https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-etcd-operator/504/pull-ci-openshift-cluster-etcd-operator-master-e2e-disruptive/1334306435525251072
> 
> This change assumes that all the static pods are running and can be queried using `crictl inspect` at the time of backup. Does your test somehow not have the static pods running?

This is not a new test - this is the quorum restore test that has been in place before I joined.

**Comment by retroflexer** [comment]
> This is not a new test - this is the quorum restore test that has been in place before I joined.

I understand that. But the test could be wrong if it is running without all the static pods. Before this change, it ididn't matter, but now it matters. 

I am running the tests wtih `set -x` to make sure I understand what is happening.

**Comment by retroflexer** [comment]
> overall the approach seems reasonable, please confirm jq status for OCP 4.4+ to ensure there is no corner cases.

Verified that 4.4.31 has `jq` installed on  their node. Using a cluster-bot on 4.4, ran this command on one of the nodes:
```
[core@ip-10-0-163-43 ~]$ sudo crictl inspect $(sudo crictl ps -q --name etcd$) | jq -r '.status.mounts | .[] | select (.containerPath=="/etc/kubernetes/static-pod-resources") | .hostPath'
/etc/kubernetes/static-pod-resources/etcd-pod-2
```
Works correctly.

**Comment by retroflexer** [comment]
Debugging locally

/hold

**Comment by hexfusion** [comment]
> Verified that 4.4.31 has jq installed on their node. Using a cluster-bot on 4.4, ran this command on one of the nodes:

Please run this by #forum-coreos so that we understand any gotchas.

**Comment by retroflexer** [comment]
Reworked the logic to get resource directory from manifests instead of using `crictl`.

/hold cancel

**Comment by retroflexer** [comment]
> Please run this by #forum-coreos so that we understand any gotchas.

Yes. They okayed the use of `jq`.

File: bindata/etcd/cluster-backup.sh [hunk 2]
```diff
@@ -34,10 +34,14 @@ function backup_latest_kube_static_resources {
...
   LATEST_RESOURCE_DIRS=()
   for RESOURCE in "${RESOURCES[@]}"; do
-    # shellcheck disable=SC2012
-    LATEST_RESOURCE=$(ls -trd "${CONFIG_FILE_DIR}"/static-pod-resources/"${RESOURCE}"-[0-9]* | tail -1) || true
+    if [ ! -f /etc/kubernetes/manifests/"${RESOURCE}".yaml ]; then
+      echo "error finding manifests for ${RESOURCE}. please check if it is running."
+      exit 1
+    fi
+
+    LATEST_RESOURCE=$(jq -r '.spec.volumes | .[] | select (.name=="resource-dir") | .hostPath.path' /etc/kubernetes/manifests/"${RESOURCE}".yaml) || true
```
Comment (hexfusion) [inline bindata/etcd/cluster-backup.sh:42]: just something to keep in mind when we backport only recently 4.7? is our static pod JSON on disk. I believe the change was introduced through library-go bump, have not dug very deep. But then this will fail in say 4.4 where it is YAML.

Comment (retroflexer) [reply bindata/etcd/cluster-backup.sh]: Thanks. Changed it to a simple `grep` command so that it works in all releases regardless of it the manifest files are in json or yaml format.

File: bindata/etcd/cluster-backup.sh [hunk 3]
```diff
@@ -28,18 +28,40 @@ if [ ! -d "$1" ]; then
...
+
+   export KUBECONFIG="/etc/kubernetes/static-pod-resources/kube-apiserver-certs/secrets/node-kubeconfigs/localhost.kubeconfig"
+   if [ ! -f "${KUBECONFIG}" ]; then
+      echo "Valid kubeconfig is not found in kube-apiserver-certs. Exiting!"
+      exit 1
+   fi
+
+   progressing=$(oc get co | awk -v OPERATOR=$operator '$1==OPERATOR { print $4 }')
+   if [ "$progressing" != "False" ]; then
+      echo "Currently the $operator is Progressing. The cluster is not healthy. Aborting!"
```
Comment (marun) [inline bindata/etcd/cluster-backup.sh:42]: s/The cluster is not healthy/A reliable backup requires that a rollout is not in progress/

Comment (retroflexer) [reply bindata/etcd/cluster-backup.sh]: Done. :+1:

File: bindata/etcd/cluster-backup.sh [hunk 4]
```diff
@@ -28,18 +28,40 @@ if [ ! -d "$1" ]; then
...
+
+   export KUBECONFIG="/etc/kubernetes/static-pod-resources/kube-apiserver-certs/secrets/node-kubeconfigs/localhost.kubeconfig"
+   if [ ! -f "${KUBECONFIG}" ]; then
+      echo "Valid kubeconfig is not found in kube-apiserver-certs. Exiting!"
+      exit 1
+   fi
+
+   progressing=$(oc get co | awk -v OPERATOR="$operator" '$1==OPERATOR { print $4 }')
+   if [ "$progressing" != "False" ]; then
+      echo "Currently the $operator is Progressing. A reliable backup requires that a rollout is not in progress.  Aborting!"
```
Comment (marun) [inline bindata/etcd/cluster-backup.sh:42]: s/is Progressing/operator is Progressing/

Comment (retroflexer) [reply bindata/etcd/cluster-backup.sh:42]: I will push again. Perhaps I shouldn't capitalize progressing, unless I put in double quotes as \"Progressing\".

File: bindata/etcd/cluster-backup.sh [hunk 5]
```diff
@@ -28,18 +28,40 @@ if [ ! -d "$1" ]; then
...
+function check_if_operator_is_progressing {
+   operator="$1"
+
+   export KUBECONFIG="/etc/kubernetes/static-pod-resources/kube-apiserver-certs/secrets/node-kubeconfigs/localhost.kubeconfig"
+   if [ ! -f "${KUBECONFIG}" ]; then
+      echo "Valid kubeconfig is not found in kube-apiserver-certs. Exiting!"
+      exit 1
+   fi
+
+   progressing=$(oc get co | awk -v OPERATOR="$operator" '$1==OPERATOR { print $4 }')
```
Comment (marun) [inline bindata/etcd/cluster-backup.sh:40]: Sorry, one more: 

`oc get co "${operator}" -o jsonpath='{.status.conditions[?(@.type=="Progressing")].status}'`

**Comment by marun** [comment]
LGTM
